ParadigmFoundation/kosu-monorepo
# go-kosu: implement validator removal and subcommand

## Overview
Implements validator removal and `kosu-cli tx update-validator sub-command. Useful to test this scenario.

## Description
Given an incoming Validator Tx we check the amount: 
	- if == 0: use the corresponding validator's power to confirm the tx and update the balance with 0
    - if != 0: use the tx amount as confirmation power.

validators updated with 0 balance will be removed from the store

## Testing
To test this, a new command has been introduced: `kosu-cli tx update-validator <block> <pubkey> <address> <power> [flags]`:

0. start the testnet
```bash
make testnet
```

1. check the current validator set:
```bash
curl localhost:8000/validators | jq '.result.validators[].voting_power'
```

2. Update a validator power
```bash
./kosu-cli --abci http://localhost:8000 --home ./testnet/node0 tx update-validator 1 r9dq+UvR+DzAmKhuo8VTk+i4OuSLA4uUPxVmKI1ucas= address 6000000000000000000
```

3. check the current validator set
```bash
curl localhost:8000/validators | jq '.result.validators[].voting_power'
```

4. Remove the validator from set
```bash
./kosu-cli --abci http://localhost:8000 --home ./testnet/node0 tx update-validator 2 r9dq+UvR+DzAmKhuo8VTk+i4OuSLA4uUPxVmKI1ucas= address 0
```

5. check the current validator set

```bash
curl localhost:8000/validators | jq '.result.validators[].voting_power'
```
